rclpy/rclpy/callback_groups.py

```diff
@@ -0,0 +1,107 @@
+# Copyright 2017 Open Source Robotics Foundation, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from threading import Lock
+import weakref
+
+
+class CallbackGroup:
+    """Control when callbacks are allowed to be executed."""
+
+    def __init__(self):
+        super().__init__()
+        self.entities = set()
+
+    def add_entity(self, entity):
+        """
+        Add an entity to the callback group.
+
+        entity - a subscription, timer, client, or service instance
+        """
+        self.entities.add(weakref.ref(entity))
+
+    def has_entity(self, entity):
+        """
+        Determine if an entity has been added to this group.
+
+        entity - a subscription, timer, client, or service instance
+        """
+        return weakref.ref(entity) in self.entities
+
+    def can_take(self, entity):
+        """
+        Return true if an entity can be executed.
+
+        entity - a subscription, timer, client, or service instance
+        """
+        raise NotImplementedError("Derived classes must implement this")
+
+    def take(self, entity):
+        """
+        Take the callback from the group to begin executing it.
+
+        entity - an entity that can be taken
+
+        Return true if the callback can be executed, false otherwise. If this returns True then
+        the method give_back must be called after the callback has been executed.
+        """
+        raise NotImplementedError("Derived classes must implement this")
+
+    def give_back(self, entity):
+        """
+        Return a callback back to the group.
+
+        entity - an entity that has been taken
+        """
+        raise NotImplementedError("Derived classes must implement this")
+
+
+class ReentrantCallbackGroup(CallbackGroup):
+    """Allow callbacks to be executed in parallel without restriction."""
+
+    def can_take(self, entity):
+        return True
+
+    def take(self, entity):
+        return True
+
+    def give_back(self, entity):
+        pass
+
+
+class MutuallyExclusiveCallbackGroup(CallbackGroup):
+    """Allow only one callback to be executing at a time."""
+
+    def __init__(self):
+        super().__init__()
+        self._active_entity = None
+        self._lock = Lock()
+
+    def can_take(self, entity):
+        with self._lock:
+            assert weakref.ref(entity) in self.entities
+            return self._active_entity is None
+
+    def take(self, entity):
```

Comment:
I agree that `take`/`give_back` doesn't really describe what's happening here, but acquire/release is closer. It's almost like the executor is borrowing the entity with `take` being "please may I borrow this" and `give_back` being "thanks, here it is". I'll update it to `can_acquire`/`acquire_for_taking`/`release_for_taking` unless something more representative comes up.

Reply:
@wjwwood How about `can_execute`, `begin_execution`/`end_execution`?

Reply:
`can_execute`, `begin_execution`, `end_execution` in b1f882ec13bc182be2a23fdf98c1296492a69756. I'm not convinced those are the right names, but I think it's an improvement. The base `Executor` still has internal methods that begin with `_take` where it matches `rcl`.

Reply:
@sloretz Those are better I think, though I would have said `can_execute_from` (meaning can execute something from this callback group), `begin_execution` or `begin_execution_of_entity` (meaning begin execution of an entity from this callback group), and `end_execution` or `end_execution_of_entity` (meaning the same as the last one), as you're not executing the callback group itself but rather something from within it. That being said, the ones I just listed with a `_from` suffix are also a bit confusing without some context. 

I'm fine with what you've got there now, with the small issue that it sounds like you're executing the callback group when you are not.

Reply:
@wjwwood Yeah, I have the same concern about the callback group is doing the execution. I think it could be improved by adding "ing" `beginning_execution`/`ending_execution`.